### DISCUSSION TITLE
Configuration of Fybrik Module Plugins

### DISCUSSION THREAD

**Discussion by cdoron** [fybrik/fybrik#956] [Developers]
FybrikModule plugins add functionality to FybrikModules. As an example, consider the arrow-flight-module which could be enhanced by plugins that perform different types of data transformations.

In this discussion we:

1. List four different options for plugin management, with emphasis on plugin configuration.
2. **We introduce the idea of registering plugins as catalog assets.**

Plugin Configuration includes: the location of the plugin (in an OCI registry), the action it implements, and credentials needed to access the registry.

The options:

1. The plugin is represented by a FybrikModule plugin CR (separate from the CR for the FybrikModule it enhances), which contains its configuration (this is the current design)
2. No separate plugin CR. The plugin appears as part of the configuration of the FybrikModule it enhances
3. The plugin is registered as a catalog asset. The asset includes all the plugin configuration. A reference to the plugin asset appears as part of the configuration of the FybrikModule it enhances
4. The final option is similar to the first in that we have a separate CR for the FybrikModule and the FybrikModule plugin. However, rather that the plugin CR containing the plugin configuration, it instead contains just the asset-ID which references the plugin configuration asset

The option we choose has no effect on the structure of the blueprint. It only affects the way that the blueprint is constructed.

**Comment by shlomitk1** [comment]
The chosen option has also effect on Fybrik configuration/deployment. Options 3 and 4 register the plugin in a catalog. Which catalog? I assume, it is not necessarily the same catalog that is used for data assets.

**Comment by simanadler** [comment]
> FybrikModule plugins add functionality to FybrikModules. As an example, consider the arrow-flight-module which could be enhanced by plugins that perform different types of data transformations.
> 
> In this discussion we:
> 
>     1. List four different options for plugin management, with emphasis on plugin configuration.
> 
>     2. **We introduce the idea of registering plugins as catalog assets.**
> 
> 
> Plugin Configuration includes: the location of the plugin (in an OCI registry), the action it implements, and credentials needed to access the registry.
> 
> The options:
> 
>     1. The plugin is represented by a FybrikModule plugin CR (separate from the CR for the FybrikModule it enhances), which contains its configuration (this is the current design)

This design was already implemented as part of the Smart Blueprint work.  FybrikModules provide the ability to declare what type of plugins they support, and a FybrikModule can be declared to be of type plugin.  So, this option is already implemented in today's code.
> 
>     2. No separate plugin CR. The plugin appears as part of the configuration of the FybrikModule it enhances

This is also already supported today.  A FybrikModule can indicate the actions it supports built in.  In this case it may or may not be implemented as a plugin, but that is transparent to the fybrik control plane and those deploying the module.
> 
>     3. The plugin is registered as a catalog asset. The asset includes all the plugin configuration. A reference to the plugin asset appears as part of the configuration of the FybrikModule it enhances

I assume you mean the data catalog.  If so, why would you put this information there?  It's not a data asset.   If you did though, then why put only plugin information in the catalog?  Why not put the modules too?

Actors that manage/use assets in a catalog are not the same actors who define/deploy modules or plugins.  Personally I don't see the business logic of putting plugin information in a data catalog.

> 
>     4. The final option is similar to the first in that we have a separate CR for the FybrikModule and the FybrikModule plugin. However, rather that the plugin CR containing the plugin configuration, it instead contains just the asset-ID which references the plugin configuration asset

Same thoughts as #3
> 
> 
> The option we choose has no effect on the structure of the blueprint. It only affects the way that the blueprint is constructed.

**Reply by roee88** [reply]
@simanadler is there any design document you can share with @cdoron that describes what is in the code today?

**Comment by roee88** [comment]
Depending on the specific plugin type, a plugin is either:
1. a deployable that affects a service, or
2. a configuration to a service

In both cases my understanding of what's implemented is that a FybrikModule of a plugin type is declared, its type matches a type declared in a non-plugin (i.e. service) FybrikModule so they could be picked up together. The plugin module declares the actions that it supports and this information is used by the application controller to facilitate building a plotter. (Note that the last part is not yet implemented AFAIK).

That's pretty good for case 1 after it's fully implemented and may also be quite good for case 2 after https://github.com/fybrik/fybrik/discussions/943 is fully implemented. But only if we truly consider these two cases while implementing. It's worth discussing the best approach for these cases now to avoid doing it wrong later. It's also worth checking if there are things that we will need to implement according to existing plans, that if we do first will cover some cases early rather than waiting for a full implementation. 

---

To focus the discussion let's start with case 2, i.e., a plugin is just some configuration to a service (existing or not). 
To make this concrete let's use the WASM prototype in arrow-data-plane. In order to use the plugin, it is required that some configuration will become available to an Apache Arrow Flight server. The configuration may include:
1. The action that the WASM module implements
2. A link to the WASM module itself (in an OCI registry)
3. A set of credentials required for downloading the WASM module, or a method for the server to obtain such credentials.
4. Arbitrary configuration required by the specific WASM module

With that in mind:

> The plugin is represented by a FybrikModule plugin CR (separate from the CR for the FybrikModule it enhances), which contains its configuration (this is the current design)

The current design allows declaring actions but not attaching configuration to them, at least not directly. A discussion is required for determining if such configuration is tied to a plugin-type FybrikModule or to an individual action within it (should a plugin module be able to support more than one action?). Overall, a discussion is required for how to add such configuration to a plugin-type FybrikModule, considering both the current implementation and the rational in #943. If this is the path we take then I hope for a clean solution that applies not just to the plugins use case but rather to configuration in general.

> No separate plugin CR. The plugin appears as part of the configuration of the FybrikModule it enhances

If I understand correctly, this approach suggests that configuration can be attached to the server FybrikModule directly, thus enabling this plugins use case sooner that having a full implementation in the manager.  The obvious downside of the approach is that we can't reuse the plugin related information and supported actions in multiple modules and that's what the previous approach enables. 

This is already technically possible today by passing the plugin info as part of `chart.values` in the FybrikModule yaml as a deployment parameter. It is equivalent to having the configuration coded into the deployment YAML directly. However, looking forward for separation of deployment from config it seems like the requirement here is to allow adding arbitrary config values similar to how we allow adding arbitrary deployment values today. I think that it's a reasonable thing to add.

> The plugin is registered as a catalog asset. The asset includes all the plugin configuration. A reference to the plugin asset appears as part of the configuration of the FybrikModule it enhances

I agree with the other comments that not all catalogs can hold such metadata. Specifically, we model our requirement as a data catalog and while catalogs out there may support keeping code, binaries, or whatever as an asset just like any other data asset - it's a requirement that we can't assume unless we go back to supporting multiple catalogs which seems undesirable. However, it's more than worth noting the rational behind your proposal as you described it to me offline because it highlights important things that we should consider in a solution. 

I initially thought that this builds upon the previous suggestion to solve the reuse problem. However, @cdoron told me that the original idea came from the credentials part of the plugin configuration: `A set of credentials required for downloading the WASM module, or a method for the server to obtain such credentials.`. Passing credentials to a module is hard, but we already have a solution for that when working with data assets - we provide a vault link to the module. If the method to use vault here too is adopted then we need a unified mechanism for working with it. From the module's code perspective it's easy because it is the same code that modules use today as long as we describe it in a similar structure.  However, eventually we will also need to setup vault policies to allow specific modules to access specific secrets. That is a responsibility of the manager and it can only be done cleanly for something structured that the manager is aware of (i.e., not just pass-through config). This is where treating it like an asset is valuable, though it doesn't mean that it really needs to be a catalog asset. Alternative solutions for passing credentials can also be designed but at this point I don't think that we want to go there. 

Other probably much less important questions raised by this suggestion are:
1. Can we really model a configuration/plugin module using the same structure as a server FybrikModule? If we can then it doesn't really matter whether we should or not at this point. But we should be sure that we can.
2. Are there other cases where it's actually useful to pass to the module metadata from the catalog for things that are not listed in the FybrikModule? Should it be a feature to consider? I currently don't see any use case for it apart from what's already discussed here.

---

Finally, the other case of plugins should also be considered: a deployable that affects a service. To make this concrete let's give an example from Istio. An `EnvoyFilter` custom resource can describe a WASM plugin to Istio sidecars. It needs to be deployed to the cluster in order to take affect (it's picked up by istiod and handled internally by Istio from there). The linking between an EnvoyFilter and a set of sidecars that it should be applied to is done by matching namespace and labels. I don't have details on if and how credentials for downloading the WASM module are supported. This use case should also be supported.
